---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_knit$set(root.dir = "D:/001_Projects/010_CONQUER_TEST/Test/")
```

# COmprehend fuNctional conseQUencEs R <img src="inst/logo/CONQUER.png" align="right" width="120" />

*Gerard Bouland, Joline Beulens, Joey Nap, Arno van der Slik, Arnaud Zaldumbide, Leen â€™t Hart and Roderick Slieker*

## Installation

Install the the development version from GitLab:  
`install.packages("devtools")`

```{r, eval = FALSE}
devtools::install_github("roderickslieker/CONQUER.DB")
devtools::install_github("roderickslieker/CONQUER.d3")
devtools::install_github("roderickslieker/CONQUER")
```


## Overview 

With the use of two functions, SNPs are summarised and visualised, namely: `summarise()` and `visualise()`.

* The `summarise()` function is used to collect all data related to SNPs. 
* The `visualise()` function initiates a RStudio Shiny-based dashboard that visualises all relevant plots. 

Note: We use the LD data from the API of NIH. You will need to register on the site to obtain a token. Please see: 

https://ldlink.nci.nih.gov/?tab=apiaccess

The token is send by email and can be provided as character string. 


## Example without multianalyze

```{r, message=F, eval=F}
library(CONQUER)

summarize(variants = c("rs878521","rs10830963"),
          directory=getwd(),
          multiAnalyze=FALSE,
          token="sometoken",
          tissues=NULL)
```

## Example with multianalyze

The available tissues can be viewed with the following command:

```{r, eval=F}
tissues <- conquer.db::gtexTissuesV8
```

```{r, message=F, eval=F}
library(CONQUER)
snps <- c("rs11642430","rs11820019","rs11842871","rs13426680","rs1377807","rs1783541",
"rs1801212","rs1801645","rs2268078","rs2581787","rs34855406","rs3802177",
"rs3810291","rs4148856","rs5213","rs6011155","rs601945","rs75423501",
"rs8010382","rs8046545")

CONQUER::summarize(variants = snps,
          directory="D:/001_Projects/010_CONQUER_TEST/Test",
          multiAnalyze=TRUE,
          token=NULL,
          tissues=c("Pancreas","Muscle_Skeletal","Liver"))


visualize(directory = "D:/001_Projects/010_CONQUER_TEST/Test", SNPs = snps)
```

## Modules pancreas 

<img src="D:/001_Projects/010_CONQUER_TEST/Modules_pancreas.png" align="center" width="500" />

&nbsp;

## Enrichment 

<img src="D:/001_Projects/010_CONQUER_TEST/Enrichment_all.png" align="center" width="500" />

&nbsp;

## LD
<img src="D:/001_Projects/010_CONQUER_TEST/LD_rs11642430.png" align="center" width="500" />

&nbsp;

## Chromosomal interactions 
<img src="D:/001_Projects/010_CONQUER_TEST/Chromosomal_interactions.png" align="center" width="500" />

&nbsp;

## Chromatin states 
<img src="D:/001_Projects/010_CONQUER_TEST/Chromatin_states.png" align="center" width="500" />

&nbsp;

## eQTLs 
<img src="D:/001_Projects/010_CONQUER_TEST/eQTLs.png" align="center" width="500" />
&nbsp;

## Gene expression 
<img src="D:/001_Projects/010_CONQUER_TEST/Expression.png" align="center" width="500" />
&nbsp;


&nbsp;
&nbsp;
&nbsp;

## Data resources

```{r, echo=F, message=F, warning=F}
resources <- CONQUER::resources
resources[is.na(resources)] <- " "
knitr::kable(x = resources)
```
